src/basic/memstream-util.c

```diff
@@ -0,0 +1,73 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "alloc-util.h"
+#include "fd-util.h"
+#include "fileio.h"
+#include "log.h"
+#include "memstream-util.h"
+
+void memstream_done(MemStream *m) {
+        assert(m);
+
+        /* First, close file stream, as the buffer may be reallocated on close. */
+        safe_fclose(m->f);
+
+        /* Then, free buffer. */
+        free(m->buf);
+}
+
+FILE* memstream_init(MemStream *m) {
+        assert(m);
+        assert(!m->f);
+
+        m->f = open_memstream_unlocked(&m->buf, &m->sz);
+        return m->f;
+}
+
+int memstream_finalize(MemStream *m, char **ret_buf, size_t *ret_size) {
+        int r;
+
+        assert(m);
+        assert(m->f);
+        assert(ret_buf);
+
+        /* Add terminating NUL, so that the output buffer is a valid string. */
+        fputc('\0', m->f);
+
+        r = fflush_and_check(m->f);
+        if (r < 0)
+                return r;
+
+        m->f = safe_fclose(m->f);
+
+        /* On fclose(), the buffer may be reallocated, and may trigger OOM. */
+        if (!m->buf)
+                return -ENOMEM;
+
```

Comment:
maybe add `assert(m->sz > 0);` here?

given the fputc() we know it must be larger than 0 here if both fflush() and fclose() worked. and given you'll subtract 1 off m->sz a bit later we should definitely check this first.